### PR TITLE
TS-37604 Made the line coverage collection less memory intense

### PR DESCRIPTION
**Pull request by stahlbauer**
Addresses issue [TS-37604](https://cqse.atlassian.net/browse/TS-37604)

- [ ] No findings in [Teamscale](https://demo.teamscale.com/activity.html#/teamscale-javascript-profiler/) (c.f. vote of the Teamscale Bot). Flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.
- [ ] Changes are tested adequately
- [ ] Profilers's README.md and the Teamscale documentation are updated in case of user-visible changes
- [ ] CHANGELOG.md of the affected packages are updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)

### CODE REVIEW

File: packages/lib-instrument/src/visitor.ts [hunk 1]
```diff
@@ -295,14 +295,22 @@ function newLineCoverageExpression(
...
+        { type: 'Identifier', name: originFileId } as Identifier,
+        { type: 'NumericLiteral', value: range.start.line } as NumericLiteral,
+        { type: 'NumericLiteral', value: range.end.line } as NumericLiteral,
+    ];
+
+    if (range.start.line === range.end.line) {
+        // Only pass one position argument if they would be the same.
+        // See also https://v8.dev/blog/adaptor-frame for performance considerations
+        argumentList = argumentList.slice(0, argumentList.length-1);
+    }
```
Comment (DreierF) [inline packages/lib-instrument/src/visitor.ts:308]: I would have a tendency to reverse the logic to:
```suggestion
    let argumentList = [
        { type: 'Identifier', name: originFileId } as Identifier,
        { type: 'NumericLiteral', value: range.start.line } as NumericLiteral
    ];

    // Only pass end line argument if they are different.
    // See also https://v8.dev/blog/adaptor-frame for performance considerations
    if (range.start.line !== range.end.line) {
        argumentList.push({ type: 'NumericLiteral', value: range.end.line } as NumericLiteral)
    }
```

File: packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts [hunk 2]
```diff
@@ -36,13 +36,24 @@ export function createCoverageBuffer(flushAfterMillis: number, onFlush: FlushFun
...
 
-        fileBuffer = { lines: [] };
+        fileBuffer = { lines: new Set(), };
         buffer.set(fileId, fileBuffer);
         return fileBuffer;
     }
 
-    function putLineCoverage(fileId: string, startLine: number, endLine: number): void {
-        getBufferFor(fileId).lines.push(startLine, endLine);
+    function putLineCoverage(fileId: string, startLine: number, endLine?: number): void {
```
Comment (DreierF) [inline packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts:44]: Not 100% sure, but this should also work:

```suggestion
    function putLineCoverage(fileId: string, startLine: number, endLine = startLine): void {
```

File: packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts [hunk 3]
```diff
@@ -36,13 +36,24 @@ export function createCoverageBuffer(flushAfterMillis: number, onFlush: FlushFun
...
+        const bufferedLines = getBufferFor(fileId).lines;
+        if (endLine) {
+            for (let line=startLine; line<endLine; line++) {
+                bufferedLines.add(line);
+            }
+        } else {
+            bufferedLines.add(startLine);
+        }
+
+        if (bufferedLines.size > 1000) {
```
Comment (DreierF) [inline packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts:54]: Wouldn't this cause coverage to take very long to get to the collector, when not much is happening on the page? Not sure how big of a problem that is in practice, but it looks like a tripping hazard to me.

**Review by DreierF — CHANGES_REQUESTED** [review]

File: packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts (hunk 3 above)
Comment (stahlbauer) [reply packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts:54]: There is an additional timer. So this is only to make sure that the buffer does not get too filled.

Comment (stahlbauer) [reply packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts:54]: Changed it to 100

Comment (DreierF) [reply packages/teamscale-javascript-instrumenter/src/vaccine/buffer.ts:54]: 💡

**Pull request merged**